packages/NodeTypeResolver/NodeTypeResolver.php

```diff
@@ -235,12 +236,29 @@ public function getNativeType(Expr $expr): Type
             }
         }
 
-        /**
-         * ArrayDimFetch got type by its Scope->getType(), otherwise, it got MixedType
-         */
-        $type = $expr instanceof ArrayDimFetch
-            ? $scope->getType($expr)
-            : $scope->getNativeType($expr);
+        $type = $scope->getNativeType($expr);
+        if ($expr instanceof ArrayDimFetch) {
+            /**
+             * Allow pull type from
+             *
+             *      - native function
+             *      - always defined by assignment
+             *
+             * eg:
+             *
+             *  $parts = parse_url($url);
```

Comment:
one more question: e.g. `pathinfo()` returns a shape like `'array{dirname?: string, basename: string, extension?: string, filename: string}'`

do we have a test-case which covers the part of optional keys (like `dirname` in this case)?

Reply:
Yes, if it checked with `isset()`, it will then be string, see example on `parse_url()` 

https://github.com/rectorphp/rector-src/blob/8ba3d90ba430782c8e9bba135411260df802df3e/tests/Issues/EmptyBooleanCompare/Fixture/fixture.php.inc#L3-L37

otherwise, that will allow empty.

Reply:
see https://github.com/rectorphp/rector-src/pull/5060